src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/base/SalesforceBaseBatchSource.java

```diff
@@ -0,0 +1,148 @@
+    Map<String, String> record = input.getValue();
+    try {
+      Schema schema = getRecordSchema(input.getKey());
+      StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+
+      for (Map.Entry<String, String> entry : record.entrySet()) {
+        String fieldName = entry.getKey();
+        String value = entry.getValue();
+
+        Schema.Field field = schema.getField(fieldName, true);
+
+        if (field == null) {
+          continue; // this field is not in schema
+        }
+
+        builder.set(field.getName(), convertValue(value, field));
+      }
+
+      emitter.emit(builder.build());
+    } catch (Exception ex) {
+      switch (getConfig().getErrorHandling()) {
+        case SKIP:
+          LOG.warn("Cannot process Salesforce row '{}', skipping it.", record, ex);
+          break;
+        case SEND:
+          StructuredRecord.Builder builder = StructuredRecord.builder(ERROR_SCHEMA);
+          builder.set(ERROR_SCHEMA_BODY_PROPERTY, record);
+          emitter.emitError(new InvalidEntry<>(400, ex.getMessage(), builder.build()));
+          break;
+        case STOP:
+          throw ex;
+        default:
+          throw new UnexpectedFormatException(
+            String.format("Unknown error handling strategy '%s'", getConfig().getErrorHandling()));
+      }
+    }
+  }
+
+  protected abstract SalesforceBaseSourceConfig getConfig();
+
+  protected abstract Schema getRecordSchema(T key);
+
+  private Object convertValue(String value, Schema.Field field) {
+    Schema fieldSchema = field.getSchema();
+
+    if (fieldSchema.isNullable()) {
+      fieldSchema = fieldSchema.getNonNullable();
+    }
+
+    Schema.Type fieldSchemaType = fieldSchema.getType();
+
+    // empty string is considered null in csv
+    if (Strings.isNullOrEmpty(value)) {
+      return null;
+    }
+
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+    if (fieldSchema.getLogicalType() != null) {
+      switch (logicalType) {
+        case DATE:
+          // date will be in yyyy-mm-dd format
+          return Math.toIntExact(LocalDate.parse(value).toEpochDay());
+        case TIMESTAMP_MICROS:
+          return TimeUnit.MILLISECONDS.toMicros(Instant.parse(value).toEpochMilli());
+        case TIME_MICROS:
+          return TimeUnit.NANOSECONDS.toMicros(LocalTime.parse(value).toNanoOfDay());
+        default:
+          throw new UnexpectedFormatException(String.format("Field '%s' is of unsupported type '%s'",
+            field.getName(), logicalType.getToken()));
+      }
+    }
+
+    switch (fieldSchemaType) {
+      case NULL:
+        return null;
+      case BOOLEAN:
+        return Boolean.parseBoolean(value);
+      case INT:
+        return Integer.parseInt(value);
+      case LONG:
+        return Long.parseLong(value);
+      case FLOAT:
+        return Float.parseFloat(value);
+      case DOUBLE:
+        return Double.parseDouble(value);
+      case BYTES:
+        return Byte.parseByte(value);
```

Comment:
this is wrong, it is parsing a single byte whereas 'bytes' is a byte array. Does salesforce have a bytes type?